File: Done/Binary_Search_307_Questions/113/Solution.py
class Solution:
    def maxValue(self, events: List[List[int]], k: int) -> int:
        n = len(events)
        events.sort(key = lambda x: x[1])
        ends = [e[1] for e in events]
        dps = []
        for i, (start, end, value) in enumerate(events):
            last_attend_index = bisect.bisect_left(ends, start) - 1
            new_dp = dps[i - 1][:] if i > 0 else [0] * k
            if value > new_dp[0]:
                    new_dp[0] = value
            if last_attend_index >= 0:
                dp = dps[last_attend_index]
                for j in range(1, k):
                    if dp[j - 1] == 0:
                        break
                    new_value = dp[j - 1] + value
                    if new_value > new_dp[j]:
                        new_dp[j] = new_value
            dps.append(new_dp)
        return max(dps[n - 1])

File: Done/Binary_Search_307_Questions/113/Solution.java
class Solution {
    public int maxValue(int[][] events, int k) {
        Arrays.sort(events, Comparator.comparingInt(a -> a[0]));
        return dfs(events, 0, -1, k);
    }

    int dfs(int[][] ev, int i, int lastEnd, int k) {
        if (i == ev.length || k == 0) return 0;
        int take = 0;
        if (ev[i][0] > lastEnd) take = ev[i][2] + dfs(ev, i + 1, ev[i][1], k - 1);
        int skip = dfs(ev, i + 1, lastEnd, k);
        return Math.max(take, skip);
    }
}